AppCenterDataStorage/AppCenterDataStorageTests/MSDataStoreTests.m

```diff
@@ -762,4 +795,152 @@ - (void)testOfflineModeBehavior {
+      XCTAssertNotNil([documentWrapper lastUpdatedDate]);
+      SomeObject *deserializedDocument = [documentWrapper deserializedValue];
+      XCTAssertNotNil(deserializedDocument);
+      XCTAssertTrue([[deserializedDocument property1] isEqualToString:@"property 1 string"]);
+      XCTAssertTrue([[deserializedDocument property2] isEqual:@42]);
+    }
+  };
+  [self waitForExpectationsWithTimeout:1 handler:handler];
+  expectation = [self expectationWithDescription:@"Get extra page"];
+  __block MSPage *testPage;
+  [testDocuments nextPageWithCompletionHandler:^(MSPage *page) {
+    testPage = page;
+    [expectation fulfill];
+  }];
+  handler = ^(NSError *_Nullable error) {
+    if (error) {
+      XCTFail(@"Expectation Failed with error: %@", error);
+    } else {
+      XCTAssertNil(testPage);
+    }
+  };
+  [self waitForExpectationsWithTimeout:1 handler:handler];
+  [httpClient stopMocking];
+}
+
+- (void)testListPagination {
+
+  // If
+  id httpClient = OCMClassMock([MSHttpClient class]);
+  OCMStub([httpClient new]).andReturn(httpClient);
+  self.sut.httpClient = httpClient;
+  id msTokenEchange = OCMClassMock([MSTokenExchange class]);
+  OCMStub([msTokenEchange retrieveCachedToken:[OCMArg any]])
+      .andReturn([[MSTokenResult alloc] initWithDictionary:[self prepareMutableDictionary]]);
+  __weak XCTestExpectation *expectation = [self expectationWithDescription:@"List first page"];
+  NSMutableDictionary *continuationHeaders = [NSMutableDictionary new];
+  continuationHeaders[@"x-ms-continuation"] = @"continuation token";
+  // First page
+  NSDictionary *firstPageHeaders = [MSCosmosDb defaultHeaderWithPartition:@"partition" dbToken:kMSTokenTest additionalHeaders:nil];
+  OCMStub([httpClient sendAsync:OCMOCK_ANY method:@"GET" headers:firstPageHeaders data:nil completionHandler:OCMOCK_ANY])
+      .andDo(^(NSInvocation *invocation) {
+        [invocation retainArguments];
+        MSHttpRequestCompletionHandler completionHandler;
+        [invocation getArgument:&completionHandler atIndex:6];
+        NSData *payload = [self getJsonFixture:@"oneDocumentPage"];
+        completionHandler(payload, [MSHttpTestUtil createMockResponseForStatusCode:200 headers:continuationHeaders], nil);
+      });
+  // Second page
+  NSDictionary *secondPageHeaders = [MSCosmosDb defaultHeaderWithPartition:@"partition"
+                                                                   dbToken:kMSTokenTest
+                                                         additionalHeaders:continuationHeaders];
+  OCMStub([httpClient sendAsync:OCMOCK_ANY method:@"GET" headers:secondPageHeaders data:nil completionHandler:OCMOCK_ANY])
+      .andDo(^(NSInvocation *invocation) {
+        [invocation retainArguments];
+        MSHttpRequestCompletionHandler completionHandler;
+        [invocation getArgument:&completionHandler atIndex:6];
+        NSData *payload = [self getJsonFixture:@"zeroDocumentsPage"];
```

Comment:
Would this ever happen?  Having a non-null continuation token and then getting a page with no documents on it?

Reply:
In theory yes. In practice it is less likely because we don't do cross-partition queries (scenario where this happens usually). For the sake of the test it's fine I think. The semantic is not wrong.

A valid example of this happening: you list a page of data and get a continuation token, at the same time someone delete the documents that were supposed to be in the next page, you use your continuation token to query the next page, it could be empty.